API redesign

bens opened the pull request:
Taking a stab at #52, I had the exact same API in mind before I saw that issue. Currently it's just a basic draft but it works for the most basic case.

bens commented on util/format.zig:
```diff
@@ -2,23 +2,10 @@ const std = @import("std");
+pub fn duration(buffer: []u8, ns: u64) ![]u8 {
+    return try std.fmt.bufPrint(buffer, "{}", .{
+        std.fmt.fmtDuration(ns),
+    });
```
An alternative to doing a hand-made formatting for durations at all, just use the stdlib.

hendriknielaender replied:
Agree!

bens commented on zbench.zig:
```diff
@@ -3,421 +3,183 @@
+            var timings_ns: std.ArrayListUnmanaged(u64) = .{};
+            var t = try std.time.Timer.start();
+            for (0..b.config.iterations) |_| {
+                b.func(arena.allocator());
```
Passing an arena here is probably a dumb idea, it just helped in getting some test code running quickly.

hendriknielaender commented on zbench.zig:
```diff
@@ -3,421 +3,183 @@
-const log = std.log.scoped(.zbench);
 
-const c = @import("./util/color.zig");
+const Color = @import("./util/color.zig").Color;
```
good adjustment, we should avoid single char vars 👍

bens commented:
AFAICT I've implemented all the features of the current library, and added a couple of things which could be moved into separate PRs to simplify review. Those are

- the incremental running API which is shown in the `progress.zig` example test,
- the `getSystemInfo` change which makes it use only stack (4KB though) and static memory so no need for an allocator,
- and the `writeJSON` functions.

I made a `Runner` abstraction which does the job of the old `run()` function but its advantage is it removes all I/O from it, making it functional and much more easily testable, you can just pass in whatever nanosecond values you want in tests and the runner's responses will be deterministic. I think this is worthwhile but it could be removed and just stick with the one-shot run function which does its own time measurement I/O.

I can take out those three changes for now but I might hold off on making PRs for them to avoid needless conflicts if some version of this PR gets merged.

I'd love a thorough review, there's obviously a lot changed but I wasn't able to make small changes to get to the desired API. Reviewing the diffs probably won't be very helpful, best just to check out the branch and play around.

bens commented:
Ok, I've moved those three features to separate branches based on this api-redesign branch, so they can wait for later I think. The diff is a little less onerous now...

hendriknielaender commented on examples/basic.zig:
```diff
@@ -21,23 +21,28 @@ fn afterAllHook() void {
+
+    const results = try bench.run();
+    defer results.deinit();
+    try results.prettyPrint(stdout, true);
```
Do we still want to give the end-user this extra line of code? For this PR, i think this is out of the scope. But my preference is that we tackle that with the  
custom output config.

https://github.com/hendriknielaender/zBench/issues/40
```zig
var bench = zbench.Benchmark.init(test_allocator, .{});
defer bench.deinit();

try bench.add("My Benchmark", myBenchmark, .{
  .iterations = 10,
  .hooks = .{
    .before_all = beforeAllHook,
    .after_all = afterAllHook,
  },
});
const results = try bench.run();
defer results.deinit();
```

bens replied:
Which extra line of code do you mean? The `results.deinit()`? I suppose it could be freed by `bench.deinit()` if we like. I tended to the results' memory being independently freed because when getting incremental results you likely want to free them after printing them out so that memory usage stays relatively constant. An alternative approach to that might be the results automatically get freed when you call the iterator again so you need to explicitly do a deep copy to keep the results around, that might be the better approach, this way just seemed more standard.

The prettyPrint line was in the old `examples/basic.zig` so I guessed you didn't mean that.
```zig
test "bench test basic" {
    const resultsAlloc = std.ArrayList(zbench.BenchmarkResult).init(test_allocator);
    var benchmarkResults = zbench.BenchmarkResults.init(resultsAlloc);
    defer benchmarkResults.deinit();
    var bench = try zbench.Benchmark.init("My Benchmark", test_allocator, .{
        .iterations = 10,
        .display_system_info = true,
        .hooks = .{
            .beforeAll = beforeAllHook,
            .afterAll = afterAllHook,
        },
    });

    try zbench.run(myBenchmark, &bench, &benchmarkResults);
    try benchmarkResults.prettyPrint();
}
```

bens replied:
Reading more closely, I think you might mean you'd like to avoid needing the `prettyPrint` line? Personally I think it's a mistake for a library to choose the writer, the client might have any reason to choose a writer other than stdout. And as far as I know there's no way to supply a writer in the `Benchmark.init()` function and store it to use it in the `run()` function since it's passed as `anytype`, but that might be my ignorance, if there is that'd be nice to pass as an optional config setting. Otherwise the simplest way to let a user choose the writer they want is just to pass it as a parameter to this `prettyPrint()` function.

Also, they might want to write the standard pretty format to stdout and write a different format to a file. IMHO, the most important thing is to give the client access to all the data they need to do whatever they need to, rather than trying to package something for every use case. If they get a `Result` value with all the timing measurements in an array then they can do anything, any kind of analysis or output format they want, that was the main need I was thinking of when writing this version of `Result`.

hendriknielaender replied:
Sorry, not to mentioned it in the first place directly that i meant `prettyPrint()`. Agree and understood your points you made. Since we have an issue discussion open about the plugin system topic. And this seems to be out of scope of this PR, lets move on.

hendriknielaender commented on util/runner/auto.zig:
```diff
@@ -0,0 +1,137 @@
+        /// before concluding. Helps in avoiding long-running benchmarks.
+        time_budget_ns: u64,
+
+        iterations_remaining: usize = 1,
```
please add comments for auto-doc for the sake of complete documentation.

bens replied:
Happy to, though neither `State` nor `State.Preparing` are exported, it's all internal.

hendriknielaender commented on zbench.zig:
```diff
@@ -34,204 +35,201 @@ pub const Config = struct {
-        p99: u64,
-        p995: u64,
-    };
+pub const BenchFunc = fn (std.mem.Allocator) void;
```
Please add auto-doc comments for at least all pub structs.

hendriknielaender reviewed (commented):
Hey @bens, first of all really cool PR. Really appreciate that you tackled this 💪. This API looks already a lot cleaner.


For the upcoming PR, especially for the json output, i will open an issue and then you can have a look if you agree with the feature request.

hendriknielaender commented on util/platform.zig:
```diff
@@ -8,11 +8,37 @@ const mac = @import("os/osx.zig");
     // ... other system information
+
+    pub fn deinit(self: OsInfo) void {
+        self.allocator.free(self.cpu);
```
When running the basic example i get an allocation error. It indicates a mismatch in the size of memory allocated versus the size of memory attempted to be freed. 

I tried it running on a macos x86_64:
`[gpa] (err): Allocation size 40 bytes does not match free size 39`

bens replied:
Thanks, I don't get that on my Debian machine so tricky to debug from here. I might see about running the examples as part of CI.

FObersteiner replied:
`[]const u8` should be a stack pointer; interesting that this is causing an allocation error. For the Linux code, I only use the allocator to dupe the string, to avoid a pointer to temporary memory, so I guess this explains why @bens doesn't get this issue on Linux. I didn't touch the OSX code since I don't have a Mac around for testing; my guess is that the call to `exec` leaves some heap memory that wants to be freed.

hendriknielaender replied:
Checked out the system-info branch with the fixed alloc 👍 this seems to resolve the error. What do you think how we should proceed? Bring this one into this PR, or as a follow up.

bens replied:
The only down side I know of for that branch is that I had to use 4KB of stack for the Windows version to not run out of space, which seems like quite a bit. It could use static memory for scratch instead, it's just that doesn't get freed afterwards, or could use different amounts for each platform. I don't know much about Windows development but from what I could find processes are [allocated 1MB by default](https://learn.microsoft.com/en-us/windows/win32/procthread/thread-stack-size), so 4KB might be fine there... It's not the sort of code that's likely to be executed in a recursion context, and after the first call it doesn't need stack again.

bens replied:
Can bring it in to this PR, it's nice to not have to debug the issue at all :smile:

hendriknielaender replied:
That would be cool 👍

FObersteiner commented on util/optional.zig:
```diff
@@ -0,0 +1,40 @@
+const std = @import("std");
+
+/// Make every field of the struct T nullable.
+pub fn Optional(comptime T: type) type {
```
that's a neat thing :)

FObersteiner reviewed (commented):
not much more to say except I like it ^^ I'm pretty much inactive at the moment when it comes to Zig 0.11, only using the 0.12-dev version for my little projects. Once 0.12 is released as "stable", I'll re-work some zBench benchmarks and will probably have more to contribute here ;-)

bens commented on util/runner/types.zig:
```diff
@@ -0,0 +1,5 @@
+const std = @import("std");
+
+pub const Error = std.mem.Allocator.Error;
```
The idea was it's any error that can be thrown from a runner, but currently that's only allocation errors.

bens commented on util/platform.zig:
```diff
@@ -30,6 +56,7 @@ pub fn linux(allocator: std.mem.Allocator) !OsInfo {
     const memory = try lnx.getTotalMemory();
 
     return OsInfo{
+        .allocator = allocator,
```
I had some other ideas around the `OsInfo` thing but moved them to another branch, this was a small change to make the structure usable as not just an internal thing. See https://github.com/bens/zBench/commit/59f8d7ae0a892bc995eba187cec4fb283310061e for the other ideas, basically I wanted to avoid needing an allocator, only use stack while fetching the info and then cache it in static memory. Then it never needs to be deinited.

hendriknielaender commented:
One last request, can you update the README. Since the config changes a bit 👍

bens commented on zbench.zig:
```diff
@@ -34,204 +35,217 @@ pub const Config = struct {
-        p995: u64,
-    };
+/// A function pointer type that represents a benchmark function.
+pub const BenchFunc = *const fn (std.mem.Allocator) void;
```
I was thinking of leaving the `Allocator` parameter for benchmark functions since it could be a relatively easy way of measuring memory use, make a allocator wrapper which records allocations and frees and pass that to the benchmark functions. It might need more thought than that but adding it now reduces API churn if it sounds like a workable idea.
